### ISSUE TITLE
Improved seeding?

### ISSUE THREAD

**Issue by cnk113** [lastz/lastz#25]
Hello,

I don't know if you are accepting suggestions for improvements, but I was wondering if you were perhaps open to implementing [geometric hashing](https://www.biorxiv.org/content/10.1101/2020.05.01.072520v1.full.pdf) for the seeding step in lastz?

**Comment by rsharris** [comment]
I saw that recent paper and had given it some thought.

I think the best way to integrate this technique, at least in the short term, would be to implement it as an external seeding stage (i.e. as a separate program). That would create a anchor segments file which would be fed into lastz (with --segments option) to perform the gapped extensions of the seeds.

The rest of this reply is me thinking about how this would be implemented inside lastz.

As I understand the paper, first-level seed hits are first collected into diagonal bands. Then within each band we look for groups of hits that are proximal. Any hits that aren't part of a such a group are discarded (right?). That part seems like it would be easy to implement in the current framework, as a filtering step at the end of the ungapped stage.

It also looks like they are using multiple spaced seed patterns for first-level seed hits, as opposed to lastz's use of a single spaced seed pattern. I'm not sure how much of my implementation would have to change to support multiple patterns. (That code was written circa 2005, no longer fresh in my memory). IIRC for a given spaced k-mer in the query, the index presents the locations of that k-mer in the reference, in reference sequence order, and this fact is made use of to prevent overlapping ungapped extensions. To implement the union of multiple spaced patterns I'd have an index for each pattern and would scan them collectively (keeping a read head for each index and taking the min position at the set of read heads). At least that would be one way to do it. So that doesn't seem too difficult.

**Comment by cnk113** [comment]
Hmm, I see perhaps tweaking the I/O in the [geometric hashing github](https://github.com/Gaius-Augustus/GeometricHashing) and feed the segments into lastz might be the easiest.

I think due to the low FP it would really help Cactus alignments from alignment slop that may be due to the improper seeding. I think this might alleviate such problems especially since many people are using external block trimmers that are over conservative, and as a result remove too much.

**Comment by rsharris** [comment]
I agree that seems like a good plan in the short term. If that proves beneficial there'd be more incentive to add support directly in lastz. It definitely seems worth exploring. But I think the last feature I added was at least five years ago, possibly longer.

As for lower FP, from my read of the paper I got the idea that that would be true of coding segments (i.e. exons) but not necessarily true for less-conserved homologies in non-coding regions.  Do you think that's correct?

There are other possible seeding schemes that might be beneficial. A few years back I was looking at building something like a spaced-seed version of a suffix array, with the idea that seeding would start with maximal spaced matches.

The caveat for all of this is the lastz project is really in maintenance mode, not ongoing development. I'm not funded to work on it (I'm funded to do other things). So while I have been doing bug fixes and responding to user help requests, I haven't done anything to improve it for several years.

**Comment by cnk113** [comment]
Sorry for the late followup, I'm slowly coming around to this. I'm slightly confused about the what stage the segment are inputted into. You mentioned it would go into the gapped extension phase, but on the docs it seems it can be used for chaining as well? In the docs, I noticed it said ```A segment file describes a list of segments representing gap-free alignments. This list is either produced internally by LASTZ as a result of the gap-free extension stage (see Overview), or read from a user-supplied file via the ‑‑segments option. The latter causes LASTZ to skip the indexing, seeding, and gap-free extension stages and begin with the chaining stage (or the next specified stage, if chaining is not requested).``` I would prefer to do chaining on my segments since I'm using the GPU based X-Drop implementation in [LOGAN](https://github.com/albertozeni/LOGAN) to create the gap-free extension. As a result I didn't do adaptive scoring so I'd really like to filter them down.

**Comment by rsharris** [comment]
It turns out that claim in the readme is incorrect (I will correct it in a moment).

As the code currently stands, the chaining stage is also skipped. In fact, the combination of --segments=xxx and --chain is rejected by the command line parser and reported as an error.

I don't recall why I made the code reject that combination. Without digging deeper, I would guess there's some data structure that chaining needed that is created during the ungapped stage that doesn't exist when segments are read from a file. ... digging ... alignment scores might not exist for segments read from the file (the score field is optional), and these are used by the chaining recurrence.

I probably also didn't envision a need to chain incoming segments, figuring that whatever created the segments would chain them. In fact, I found evidence today that I envisioned external chaining as one of the use cases for importing segments. That's not to say the need doesn't exist, just that I didn't envision it.

The possible solutions that come to mind are
(a) incorporate chaining of segments in lastz.
(b) perform chaining of segments in an external C program, making use of existing modules in lastz.
(c) implement chaining of segments as an external package in another language, e.g. python.

Option (c) seems like the easiest to me at this point. And has the advantage that other chaining algorithms could be explored.

Creating the correct data structures for (b) could be a dragon's nest. Probably more effort than (a) but has the advantage that it can't affect other use cases.

If this were 2010 (2012?) when I was still actively working on lastz, I'd opt for (a). But my familiarity with the source code as waned over the past decade. It seems like all I would need to do is (1) score the segments if scores weren't provided, and (2) call the chaining function. But finding the right place to do those operations, and properly condition their activation, would take some effort. Plus testing to make certain other use cases aren't affected.
